layabout.py

```diff
@@ -1,3 +1,287 @@
+            Initially :obj:`None`, the attribute is set after calling the
+            :obj:`Layabout.connect` or :obj:`Layabout.run`
+            methods.
+
+    Example:
+
+        .. code-block:: python
+
+           from layabout import Layabout
+
+           layabout = Layabout('app')
+
+
+           @layabout.handle('message')
+           def echo(slack, event):
+               \"\"\" Echo all messages seen by the app. \"\"\"
+               channel = event['channel']
+               message = event['text']
+               subtype = event.get('subtype')
+
+               # Avoid an infinite loop of echoing our own messages.
+               if subtype != 'bot_message':
+                   slack.rtm_send_message(channel, message)
+
+           layabout.run()
+    """
+    def __init__(self, name: str, env_var: str = 'SLACK_API_TOKEN') -> None:
+        # TODO: Consider keyword only arguments.
+        self.name = name
+        self._env_var = env_var
+        self._token: str = None
+        self.slack: SlackClient = None
+        self._handlers: _Handlers = defaultdict(list)
+
+    def handle(self, type: str, kwargs: dict = None) -> Callable:
+        """
+        Register an event handler with the :obj:`Layabout` instance.
+
+        Args:
+            type: The name of a Slack RTM API event to be handled. As a
+                special case, although it is not a proper RTM event, ``*`` may
+                be provided to handle all events. For more information about
+                available events see the
+                `Slack RTM API <https://api.slack.com/rtm>`_.
+            kwargs: Optional arbitrary keyword arguments passed to the event
+                handler when the event is triggered.
+
+        Returns:
+            A decorator that validates and registers a Layabout event handler.
+
+        Raises:
+            TypeError: If the decorated :obj:`Callable`'s signature does not
+                accept at least 2 parameters.
+        """
+        def decorator(fn: Callable) -> Callable:
+            # Validate that the wrapped callable is a suitable event handler.
+            sig = signature(fn)
+            if len(sig.parameters) < 2:
+                raise TypeError("Layabout event handlers take at least 2 "
+                                "positional arguments, a slack client and an "
+                                "event")
+
+            # Register a tuple of the callable and its kwargs, if any.
+            self._handlers[type].append((fn, kwargs or {}))
+
+            @wraps(fn)
+            def wrapper(slack: SlackClient, event: Dict[str, Any],
+                        **kwargs: Any) -> Any:
+                # We don't actually do anything with the return value, but this
+                # might make unit testing easier for users.
+                return fn(slack, event, **kwargs)
+            return wrapper
+        return decorator
+
+    def connect(self, token: str = None) -> bool:
```

Comment:
I think this should be a private function and if people want to circumvent `run`'s normal control flow they can do so at their own risk. =P

This does make me wonder if you have to worry about socket timeouts for this API. What happens if you call run 2 hours apart?

Does it also cover API token expiration?

Reply:
I think that's fair. I'm still on the fence about it, but it's good to have your perspective. If anything I'm leaning towards your opinion as well.

Socket timeouts or websocket failures can and do happen frequently, which is part of the reason for `Layabout._reconnect`. There's actually a [`reconnect_url`](https://api.slack.com/events/reconnect_url) event, but it's listed as experimental, so I'm sure that's why the [slackclient](https://github.com/slackapi/python-slackclient) library doesn't handle reconnections for us and why I'm not super comfortable relying on that either. So I just let the connections fail when I attempt to fetch events, then try reconnecting.

As for token expiration, Slack API tokens basically don't expire. I've had the same one for one of my bots for well over a year.

Reply:
This is actually a really good point though. With the `until` parameter, it's totally possible that someone would call `run` 2 hours apart. Right now if the socket was still open then it'd just get used, but if it was disconnected a reconnect attempt would happen. Perhaps the appropriate thing to do is to actually shut the socket down after `run` is done and let it get reconnected if it's called again later.